subsys/bluetooth/host/smp.c

```diff
@@ -4706,6 +4706,7 @@ static void bt_smp_encrypt_change(struct bt_l2cap_chan *chan,
 	 */
 	if (IS_ENABLED(CONFIG_BT_CENTRAL) &&
 	    IS_ENABLED(CONFIG_BT_PRIVACY) &&
+	    conn->role == BT_HCI_ROLE_CENTRAL &&
```

Comment:
Hmm, interesting that we have relied on just the Kconfig option without checking the role here. 

It looks like this issue exist and may change behavior several other places. Created https://github.com/zephyrproject-rtos/zephyr/issues/63514 to track this (generic) issue.